pkg/storage/azurecdn/storage.go

```diff
@@ -0,0 +1,82 @@
+package azurecdn
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/Azure/azure-storage-blob-go/2017-07-29/azblob"
+)
+
+// Storage implements (github.com/gomods/athens/pkg/storage).Saver and
+// also provides a function to fetch the location of a module
+type Storage struct {
+	accountURL *url.URL
+	cred       azblob.Credential
+}
+
+// New creates a new azure CDN saver
+func New(accountName, accountKey string) (*Storage, error) {
+	u, err := url.Parse(fmt.Sprintf("https://%s.blob.core.windows.net", accountName))
+	if err != nil {
+		return nil, err
+	}
+	cred := azblob.NewSharedKeyCredential(accountName, accountKey)
+	return &Storage{accountURL: u, cred: cred}, nil
+}
+
+// BaseURL returns the base URL that stores all modules. It can be used
+// in the "meta" tag redirect response to vgo.
+//
+// For example:
+//
+//	<meta name="go-import" content="gomods.com/athens mod BaseURL()">
+func (s Storage) BaseURL() *url.URL {
+	return s.accountURL
+}
+
+// Save implements the (github.com/gomods/athens/pkg/storage).Saver interface.
+func (s *Storage) Save(module, version string, mod, zip, info []byte) error {
+	ctx := context.Background()
+
+	pipe := azblob.NewPipeline(s.cred, azblob.PipelineOptions{})
+	serviceURL := azblob.NewServiceURL(*s.accountURL, pipe)
+	// rules on container names:
+	// https://docs.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata#container-names
+	//
+	// This container must exist
+	containerURL := serviceURL.NewContainerURL("gomodules")
+
+	infoBlobURL := containerURL.NewBlockBlobURL(fmt.Sprintf("%s/@v/%s.info", module, version))
+	modBlobURL := containerURL.NewBlockBlobURL(fmt.Sprintf("%s/@v/%s.mod", module, version))
+	zipBlobURL := containerURL.NewBlockBlobURL(fmt.Sprintf("%s/@v/%s.zip", module, version))
```

Comment:
Between [the formerly referenced naming rules](https://docs.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata#blob-names) and [RFC 3986 pg. 12](http://www.ietf.org/rfc/rfc3986.txt), I _believe_ you'll need to escape the '@' sign here.

Reply:
@marstr good point. I was curious about this so I tried it out and [it works](https://testvgo.blob.core.windows.net/gomodules/swtch.com/testmod/@v/list). Did I miss something?

Reply:
@marstr when I test this out, I'll report back if it really works or not

Reply:
If it works, it works! Something in the Blob library must be escaping it for you, or the documentation is just inaccurate.